pyamg/amg_core/ruge_stuben.h

```diff
@@ -956,345 +943,1339 @@ void cr_helper(const I A_rowptr[],  const int A_rowptr_size,
+                }
+            }
+
+            // Set entries in P (interpolation weights w_ij from strongly connected C-points)
+            I nnz = Pp[i];
+            for (I jj = Sp[i]; jj < Sp[i+1]; jj++) {
+
+                if (splitting[Sj[jj]] == C_NODE) {
+
+                    // Set temporary value for Pj as global index, j. Will be mapped to
+                    // appropriate coarse-grid column index after all data is filled in. 
+                    Pj[nnz] = Sj[jj];
+                    I j = Sj[jj];
+
+                    // Initialize numerator as a_ij
+                    T numerator = Sx[jj];
+
+                    // Sum over strongly connected fine points
+                    for (I kk = Sp[i]; kk < Sp[i+1]; kk++) {
+                        if ( (splitting[Sj[kk]] == F_NODE) && (Sj[kk] != i) ) {
+                            
+                            // Get column k and value a_ik
+                            I k = Sj[kk];
+                            T a_ik = Sx[kk];
+
+                            // Get a_kj (have to search over k'th row in A for connection a_kj)
+                            T a_kj = 0;
+                            for (I search_ind = Ap[k]; search_ind < Ap[k+1]; search_ind++) {
+                                if ( Aj[search_ind] == j ){
+                                    a_kj = Ax[search_ind];
+                                    break;
+                                }
+                            }
+
+                            // If a_kj == 0, then we don't need to do any more work, otherwise
+                            // proceed to account for node k's contribution
+                            if (std::abs(a_kj) > 1e-16) {
+                                
+                                // Calculate sum for inner denominator (loop over strongly connected C-points)
+                                T inner_denominator = 0;
+                                for (I ll = Sp[i]; ll < Sp[i+1]; ll++) {
+                                    if (splitting[Sj[ll]] == C_NODE) {
+                                        
+                                        // Get column l
+                                        I l = Sj[ll];
+                                        
+                                        // Add connection a_kl if present in matrix (search over kth row in A for connection)
+                                        for (I search_ind = Ap[k]; search_ind < Ap[k+1]; search_ind++) {
+                                            if (Aj[search_ind] == l) {
+                                                inner_denominator += Ax[search_ind];
+                                            }
+                                        }
+                                    }
+                                }
+
+                                // Add a_ik * a_kj / inner_denominator to the numerator 
+                                if (std::abs(inner_denominator) < 1e-16) {
```

Comment:
Should this cutoff be relative in some way, say
`std::abs(inner_denominator) / std:abs(a_ik*a_kj) < 1e-16`
?